core/src/main/java/org/apache/calcite/jdbc/JavaTypeFactoryImpl.java

```diff
@@ -243,11 +243,24 @@ public RelDataType toSql(RelDataType type) {
   /** Converts a type in Java format to a SQL-oriented type. */
   public static RelDataType toSql(final RelDataTypeFactory typeFactory,
       RelDataType type) {
+    return toSql(typeFactory, type, true);
+  }
+
+  private static RelDataType toSql(final RelDataTypeFactory typeFactory,
+      RelDataType type, boolean isRootType) {
     if (type instanceof RelRecordType) {
-      return typeFactory.createStructType(
-          Lists.transform(type.getFieldList(),
-              field -> toSql(typeFactory, field.getType())),
-          type.getFieldNames());
+      RelDataType structType = typeFactory.createStructType(
+              Lists.transform(type.getFieldList(),
+                field -> toSql(typeFactory, field.getType(), false)),
+              type.getFieldNames());
+
+      // struct types are created by default with nullability false, so if we are dealing
+      // with the "root" type and it should be nullable, change its nullability
+      if (isRootType && type.isNullable()) {
```

Comment:
I would move this if at the end of the method and unify setting the nullability for struct and java types. Checking if the type is nullable is not necessary since the control is already incorporated inside `createTypeWithNullability` method.